text/0000-build-std.md

```diff
@@ -276,52 +357,1098 @@ proposed by this RFC.
+[why-replace-restricted_std-with-explicit-standard-library-support-for-a-target]: #why-replace-restricted_std-with-explicit-standard-library-support-for-a-target
+
+`restricted_std` was originally added as part of a mechanism to enable the
+standard library to build on all targets (just with stubbed out functionality),
+however stability is not an ideal match for this use case. When `restricted_std`
+applies, users must add `#![feature(restricted_std)]` to opt-in to using the
+standard library anyway (conditionally, only for affected targets), and have no
+mechanism for opting-in on behalf of their dependencies (including first-party
+crates like `libtest`).
+
+## Why disallow custom targets?
+[why-disallow-custom-targets]: #why-disallow-custom-targets
+
+While custom targets can be used on stable today, in practice, they are only
+used on nightly as `-Zbuild-std` would need to be used to build at least `core`.
+As such, if build-std were to be stabilised, custom targets would become much
+more usable on stable toolchains.
+
+In order to avoid users relying on the unstable target-spec-json format on a
+stable toolchain, using custom targets with build-std on a stable toolchain is
+disallowed until another RFC can consider all the implications of this
+thoroughly.
+
+## Why explicitly declare dependencies on the standard library in `Cargo.toml`?
+[why-explicitly-declare-dependencies-on-the-standard-library-in-cargo-toml]: #why-explicitly-declare-dependencies-on-the-standard-library-in-cargotoml
+
+If there are no explicit dependencies on standard library crates, Cargo would
+need to be able to determine which standard library crates to build when this is
+required:
+
+- Cargo could unconditionally build `std`, `alloc` and `core`. Not only would
+  this be unnecessary and wasteful for `no_std` crates in the embedded
+  ecosystem, but sometimes a target may not support building `std` at all and
+  this would cause the build to fail.
+- rustc could support a `--print` value that would print whether the crate
+  declares itself as `#![no_std]` crate, and based on this, Cargo could build
+  `std` or only `core`. Cargo would also need to know whether to build `alloc`
+  too, which checking for `#![no_std]` does not help with. Cargo could go
+  further and ask rustc whether a crate (or its dependencies) used `alloc`, but
+  this seems needlessly complicated.
+
+Furthermore, supporting explicit dependencies on standard library crates enables
+use of other Cargo features that apply to dependencies in a natural and
+intuitive way. If there were not explicit standard library dependencies and
+enabling features on the `std` crate was desirable, then a mechanism other than
+the standard syntax for this would be necessary, such as a flag (e.g.
+`-Zbuild-std-features`) or option in Cargo's configuration. This also applies to
+optional dependencies, public/private features, etc.
+
+**TODO:** document alternatives, or at least link to "why not automatic" below
```

Comment:
With the frontmatter feature (https://github.com/rust-lang/rust/issues/136889), Cargo needs to parse the entry point source file anyway so I think it's a reasonable alternative to suggest it could detect the presence of `#![no_std]`. As you say above though, the main reason not to do that is it does not enable the idiomatic use of other Cargo features.

Reply:
Added to the TODO to mention this